lib/jxl/enc_heuristics.cc

```diff
@@ -207,6 +207,125 @@ bool DefaultEncoderHeuristics::HandlesColorConversion(
+      0.024831698004347,   0.0047301162858671,  -0.031030197792468,
+      -0.031030197792468,  0.0047301162858671,  0.024831698004347,
+      0.01473598935701,    -0.0012016220643441, -0.0042216781245469,
+      -0.0047187050820992, -0.039563875032655,  0.010727545221453,
+      0.11040774078162,    0.024831698004347,   -0.19568935972668,
+      -0.19568935972668,   0.024831698004347,   0.11040774078162,
+      0.010727545221453,   -0.039563875032655,  -0.0047187050820992,
+      0.0024550503214722,  -0.0040179839838831, -0.0089432318159899,
+      0.010727545221453,   0.01473598935701,    -0.0051874044212185,
+      -0.0051874044212185, 0.01473598935701,    0.010727545221453,
+      -0.0089432318159899, -0.0040179839838831, 0.0024550503214722,
+      0.0019813989090779,  0.017461064645154,   -0.0040179839838831,
+      -0.039563875032655,  -0.0012016220643441, 0.083265651677212,
+      0.083265651677212,   -0.0012016220643441, -0.039563875032655,
+      -0.0040179839838831, 0.017461064645154,   0.0019813989090779,
+      0.017242556564832,   0.0019813989090779,  0.0024550503214722,
+      -0.0047187050820992, -0.0042216781245469, 0.0073616744679974,
+      0.0073616744679974,  -0.0042216781245469, -0.0047187050820992,
+      0.0024550503214722,  0.0019813989090779,  0.017242556564832};
+
+  int64_t xsize = input.xsize();
+  int64_t ysize = input.ysize();
+
+  for (size_t y = 0; y < output->ysize(); y++) {
+    float* row_out = output->Row(y);
+    const float* row_in[kernely];
+    // get the rows in the support
+    for (size_t ky = 0; ky < kernely; ky++) {
+      int64_t iy = y * 2 + ky - (kernely - 1) / 2;
+      if (iy < 0) iy = 0;
+      if (iy >= ysize) iy = ysize - 1;
+      row_in[ky] = input.Row(iy);
+    }
+
+    for (size_t x = 0; x < output->xsize(); x++) {
+      // get min and max values of the original image in the support
+      float min = std::numeric_limits<float>::max();
+      float max = std::numeric_limits<float>::min();
+      // kernelx - R and kernely - R are the radius of a square region in which
+      // the values of a pixel are bounded to reduce ringing.
+      static constexpr int64_t R = 5;
+      for (int64_t ky = R; ky + R < kernely; ky++) {
+        for (int64_t kx = R; kx + R < kernelx; kx++) {
+          int64_t ix = x * 2 + kx - (kernelx - 1) / 2;
+          if (ix < 0) ix = 0;
+          if (ix >= xsize) ix = xsize - 1;
+          min = std::min<float>(min, row_in[ky][ix]);
+          max = std::max<float>(max, row_in[ky][ix]);
+        }
+      }
+
+      float sum = 0;
+      for (int64_t ky = 0; ky < kernely; ky++) {
+        for (int64_t kx = 0; kx < kernelx; kx++) {
+          int64_t ix = x * 2 + kx - (kernelx - 1) / 2;
+          if (ix < 0) ix = 0;
+          if (ix >= xsize) ix = xsize - 1;
+          sum += row_in[ky][ix] * kernel[ky * kernelx + kx];
+        }
+      }
+      row_out[x] = sum;
+      if (row_out[x] < min) row_out[x] = min;
+      if (row_out[x] > max) row_out[x] = max;
```

Comment:
This looks weird and can cause artefacts -- we may need to explore in the future better approaches such as modulating the allowed range based on amplitude of changes, and local noise environment.

When clipping we may want to propagate the error into 4 connected pixels (or otherwise) so that volume is preserved. Otherwise there can be darkening/brightening in weird conditions.

Reply:
As seen on test images, indeed it does and without the clipping it looks amazing in many areas of the image, except for the large amount of ringing. Better heuristics in future pull requests will help